src/Microsoft.AspNetCore.Mvc.ApiExplorer/ApiResponseTypeProvider.cs

```diff
@@ -0,0 +1,339 @@
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Controllers;
+using Microsoft.AspNetCore.Mvc.Formatters;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+
+namespace Microsoft.AspNetCore.Mvc.ApiExplorer
+{
+    internal class ApiResponseTypeProvider
+    {
+        private readonly IModelMetadataProvider _modelMetadataProvider;
+        private readonly IActionResultTypeMapper _mapper;
+        private readonly MvcOptions _mvcOptions;
+
+        public ApiResponseTypeProvider(
+            IModelMetadataProvider modelMetadataProvider,
+            IActionResultTypeMapper mapper,
+            MvcOptions mvcOptions)
+        {
+            _modelMetadataProvider = modelMetadataProvider;
+            _mapper = mapper;
+            _mvcOptions = mvcOptions;
+        }
+
+        public IList<ApiResponseType> GetApiResponseTypes(ControllerActionDescriptor action)
+        {
+            // We only provide response info if we can figure out a type that is a user-data type.
+            // Void /Task object/IActionResult will result in no data.
+            var declaredReturnType = GetDeclaredReturnType(action);
+
+            var runtimeReturnType = GetRuntimeReturnType(declaredReturnType);
+
+            var responseMetadataAttributes = GetResponseMetadataAttributes(action);
+            if (responseMetadataAttributes.Length == 0)
+            {
+                // Action does not have any conventions. Look for conventions on the type.
+                responseMetadataAttributes = GetResponseMetadataAttributesFromConventions(action);
+            }
+
+            var apiResponseTypes = GetApiResponseTypes(responseMetadataAttributes, runtimeReturnType);
+            return apiResponseTypes;
+        }
+
+        private IApiResponseMetadataProvider[] GetResponseMetadataAttributesFromConventions(ControllerActionDescriptor action)
+        {
+            if (action.FilterDescriptors == null)
+            {
+                return Array.Empty<IApiResponseMetadataProvider>();
+            }
+
+            var filterDescriptors = action.FilterDescriptors?.Where(f => f.Filter is ApiConventionAttribute).ToArray();
+            if (filterDescriptors.Length == 0)
+            {
+                return Array.Empty<IApiResponseMetadataProvider>();
+            }
+
+            foreach (var filterDescriptor in filterDescriptors)
+            {
+                var apiConventionAttribute = (ApiConventionAttribute)filterDescriptor.Filter;
+                var method = GetConventionMethod(action.MethodInfo, apiConventionAttribute.ConventionType);
+                if (method == null)
+                {
+                    continue;
+                }
+
+                return method.GetCustomAttributes(inherit: false)
+                    .OfType<IApiResponseMetadataProvider>()
+                    .ToArray();
+            }
```

Comment:
This code here (starting with L59) could probably be simplified. For instance this does a LINQ query and then does a case. This could just be a select + oftype.